# Grouping, quartiles and ranges

We saw in our cursory review of the descriptives that there is evidence the data is skewed. We also saw that there are some other interesting patterns to the data. We saw for example that the range of scores was quite large. However, we also saw that there appeared to a large amount of intra-statistic-consistency across each of the areas measured (Open, High, Low etc.). 

Ahead of further analysis we want to understnd these patterns better, such that we better frame further tests. It might be that there are companies we want to exclude, it may be that there are specific time period that impact the data. These findings might tell us we need to transform the data for tests we will later use. We may find that the year has an impact on the data also, which will further change how we want to analyse further.

Importantly, there are 500 companies in the data with 9 columns of values daily over a 5 year period. This is equivalent to approximately **5,850,000** individual values. This poses a significant computational challenge. We therefore need to find a way to minimize this number so that we can more effectively compute. We will therefore average high and low scores for each company and use only these means. 

We will then Find the first and fourth quartiles of this data. We can then compare these score and more importantly, test to see if either the fourth or the 1st quartile differ significntly enough from the second and third quartiles.

**We will do the following:**
- Find the top quartile of companies
- Find the bottom quartile of companies
- Compare first and fourth with the second and third

In [2]:
# Import modules and variables we need 
import pandas as pd
yfin_csv = pd.read_csv(r'https://raw.githubusercontent.com/alexcrockett/Jupyter-Playground/personal/yfin_dataset/02-data/stock_details_5_years.csv')

                        Date        Open        High         Low       Close  \
0  2018-11-29 00:00:00-05:00   43.829761   43.863354   42.639594   43.083508   
1  2018-11-29 00:00:00-05:00  104.769074  105.519257  103.534595  104.636131   
2  2018-11-29 00:00:00-05:00   54.176498   55.007500   54.099998   54.729000   
3  2018-11-29 00:00:00-05:00   83.749496   84.499496   82.616501   83.678497   
4  2018-11-29 00:00:00-05:00   39.692784   40.064904   38.735195   39.037853   

      Volume  Dividends  Stock Splits Company  
0  167080000       0.00           0.0    AAPL  
1   28123200       0.00           0.0    MSFT  
2   31004000       0.00           0.0   GOOGL  
3  132264000       0.00           0.0    AMZN  
4   54917200       0.04           0.0    NVDA  


In [3]:
# Finding the fourth quartile of data

## Compute the average between high and low for each company:
yfin_csv['Average'] = (yfin_csv['High'] + yfin_csv['Low']) / 2 # Here we are defining a new column called Average

## Group by company and compute the mean of the 'Average' column
company_averages = yfin_csv.groupby('Company')['Average'].mean() # Now we create a new variable to group companies with their means and compute means for these groups

## Find the quartiles
first_quartile = company_averages.quantile(0.25) # Define the 1st quartile
fourth_quartile = company_averages.quantile(0.75) #Define the 4th quartile

## Segment the quartiles
companies_in_first_quartile = company_averages[company_averages <= first_quartile] # Identify those in the 1st quartile
companies_in_fourth_quartile = company_averages[company_averages >= fourth_quartile] # Identify those in the 4th quartile

## Print the first quartile
print("Companies in the first quartile:")
print(companies_in_first_quartile)
print("-------------------")

print("Companies in the fourth quartile:")
print(companies_in_fourth_quartile)
print("-------------------")

# Sort the data

## Sort companies in the first quartile in ascending order to find the lowest
lowest_in_first_quartile = companies_in_first_quartile.sort_values(ascending=True)
highest_in_first_quartile = companies_in_first_quartile.sort_values(ascending=False)

## Sort companies in the fourth quartile in descending order to find the highest
lowest_in_fourth_quartile = companies_in_fourth_quartile.sort_values(ascending=True)
highest_in_fourth_quartile = companies_in_fourth_quartile.sort_values(ascending=False)

## Print the lowest company in the first quartile
print("Lowest company in the first quartile:")
print(lowest_in_first_quartile.head(1))  # .head(1) gets the top company after sorting in ascending order
print("-------------------")

## Print the highest company in the first quartile
print("Highest in the first quartile")
print(highest_in_first_quartile.head(1))
print("-------------------")

## Print the lowest company in the fourth quartile
print("Lowest company in the fourth quartile:")
print(lowest_in_fourth_quartile.head(1))  # .head(1) gets the top company
print("-------------------")


# Print the highest company in the fourth quartile
print("Highest company in the fourth quartile:")
print(highest_in_fourth_quartile.head(1))  # .head(1) gets the top company
print("-------------------")


Companies in the first quartile:
Company
ABEV     2.947499
BAC     30.790636
BBD      4.368681
BBVA     4.879372
BCE     40.470195
          ...    
WFC     39.525165
WIT      5.356330
WMB     23.720061
WPM     36.569403
WY      28.026198
Name: Average, Length: 123, dtype: float64
-------------------
Companies in the fourth quartile:
Company
ACN     249.476417
ADBE    417.060667
ADP     184.081445
ADSK    215.628537
AMGN    212.491112
           ...    
VRTX    247.068470
WDAY    202.697743
WST     270.453652
WTW     204.263422
ZM      173.939238
Name: Average, Length: 123, dtype: float64
-------------------
Lowest company in the first quartile:
Company
LYG    2.102133
Name: Average, dtype: float64
-------------------
Highest in the first quartile
Company
BK    42.263321
Name: Average, dtype: float64
-------------------
Lowest company in the fourth quartile:
Company
VMC    155.851637
Name: Average, dtype: float64
-------------------
Highest company in the fourth quartile:
Company
NVR  

# Notes on medians
In the above code we created a new column and created an average, for each row, based on the highest and lowest scores. We then grouped scores by company and then took the mean for each group. This gave us a smaller set of data to work with. We then computed the 1st and 4th quartiles. 

Even though there may be some loss of fidelity in this view, it gives us a view into the companies that may be having an effect on the data.We can use this to make decisions about the original dat set.

We can see that the company with the lowest mean is LYG with 2.10 USD and the highest mean is NVR with 4369.58 USD.

In the notebook 'fin_descriptives.ipynb' we saw that the average low score was 138.28 USD and the average high score was 141.85 (variables mean_low and mean_high respectively). In other words our lowest score is  136.18 USD from the mean and the highest is 4,227.73 from mean_high. 

In fact looking at the highest value in the 1st quartile and the lowest value in the fourth quartile gives a little more color to the data. The highest value in the 1st quartile is around 6.8 times smaller than the standard deviation while lowest value in the fourth quartile is more than 1/2 the size of the standard deviation. 

Qualitatively this suggests that there is positive skew that most likely is impacting the shape. It therefore makes sense to take the follwoing steps:

- Check to see which companies have a significantly sized range
- Use the interquartile range to exlude outliers. 
- Graph the results 

In [4]:
# Range statistics

## Defining the range values
yfin_csv['Range'] = yfin_csv['High'] - yfin_csv['Low'] # Subtracting Low from High to get the actual range

## Grouping companies by their range and calculating statistics
company_ranges_mean = yfin_csv.groupby('Company')['Range'].mean()
company_ranges_min = yfin_csv.groupby('Company')['Range'].min()
company_ranges_max = yfin_csv.groupby('Company')['Range'].max()
company_ranges_std = yfin_csv.groupby('Company')['Range'].std()

## Sorting by the standard deviation of the range scores

### Mean
highest_mean_range = company_ranges_mean.sort_values(ascending=False)
lowest_mean_range = company_ranges_mean.sort_values(ascending=True)

### Min
highest_min_range = company_ranges_min.sort_values(ascending=False)
lowest_min_range = company_ranges_min.sort_values(ascending=True)

### Max
highest_max_range = company_ranges_max.sort_values(ascending=False)
lowest_max_range = company_ranges_max.sort_values(ascending=True)

### Std
highest_std_range = company_ranges_std.sort_values(ascending=False)
lowest_std_range = company_ranges_std.sort_values(ascending=True)

## Printing

### Mean scores
print("Mean Range")
print(company_ranges_mean)
print("-------------------")

### Min scores
print("Min Range")
print(company_ranges_min)
print("-------------------")

### Max scores
print("Max Range")
print(company_ranges_max)
print("-------------------")

### Standard deviation scores
print("Standard Deviation of Range")
print(company_ranges_std)
print("-------------------")

print("-------------------")
print("-------------------")

### Highest/lowest mean
print("Company with Highest Mean Range")
print(highest_mean_range.head(1))
print("-------------------")

print("Company with Lowest Mean Range")
print(lowest_mean_range.head(1))
print("-------------------")

### Highest/lowest Min
print("Company with Highest Min Range")
print(highest_min_range.head(1))
print("-------------------")

print("Company with Lowest Min Range")
print(lowest_min_range.head(1))
print("-------------------")

### Highest/lowest Max
print("Company with Highest Max Range")
print(highest_max_range.head(1))
print("-------------------")

print("Company with Lowest Max Range")
print(lowest_max_range.head(1))
print("-------------------")

### Highest/lowest Std
print("Company with Highest Std Range")
print(highest_std_range.head(1))
print("-------------------")

print("Company with Lowest Std Range")
print(lowest_std_range.head(1))
print("-------------------")


Mean Range
Company
A       2.481682
AAPL    2.666117
ABBV    2.109405
ABEV    0.082863
ABNB    6.198645
          ...   
YUM     1.918100
ZBH     2.892501
ZM      8.445569
ZS      6.485945
ZTS     3.299444
Name: Range, Length: 491, dtype: float64
-------------------
Min Range
Company
A       0.435815
AAPL    0.281952
ABBV    0.446642
ABEV    0.026536
ABNB    1.400002
          ...   
YUM     0.486592
ZBH     0.889999
ZM      0.883999
ZS      0.619999
ZTS     0.670361
Name: Range, Length: 491, dtype: float64
-------------------
Max Range
Company
A       11.565640
AAPL    12.566008
ABBV    14.103725
ABEV     0.610324
ABNB    30.500000
          ...    
YUM     10.221042
ZBH     15.164343
ZM      68.179993
ZS      59.039978
ZTS     15.160277
Name: Range, Length: 491, dtype: float64
-------------------
Standard Deviation of Range
Company
A       1.389413
AAPL    1.697801
ABBV    1.178597
ABEV    0.038495
ABNB    3.923734
          ...   
YUM     1.028575
ZBH     1.561640
ZM      8.151012
Z

In [8]:
### Highest/lowest Max
print("Company with Highest Max Range")
print(highest_max_range.head(20))
print("-------------------")

print("Company with Lowest Min Range")
print(lowest_min_range.head(20))
print("-------------------")

Company with Highest Max Range
Company
NVR      657.209961
BKNG     216.750000
AZO      194.699951
BIIB     182.549988
MELI     180.000000
CMG      135.239990
HUBS     128.965027
MTD      127.150024
AVGO     121.655790
COIN     119.540009
FCNCA    110.360107
SNOW      87.889999
FICO      87.010010
MRNA      84.096985
REGN      81.289978
BNTX      80.730696
TDG       78.132938
NOW       73.509979
ZM        68.179993
MSCI      67.473685
Name: Range, dtype: float64
-------------------
Company with Lowest Min Range
Company
CARR    0.000000
FERG    0.000000
NWG     0.000000
MBLY    0.000000
SYM     0.000000
DB      0.000000
IBKR    0.000000
BKR     0.000000
CHD     0.000000
CCI     0.000000
BIIB    0.000000
ALNY    0.000000
UMC     0.007700
LYG     0.008572
MFG     0.010000
WDS     0.015389
BBVA    0.016174
TEF     0.016342
SAN     0.018170
NOK     0.018910
Name: Range, dtype: float64
-------------------
